packages/flutter_tools/test/base/build_test.dart

```diff
@@ -0,0 +1,192 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:file/memory.dart';
+import 'package:flutter_tools/src/base/build.dart';
+import 'package:flutter_tools/src/base/file_system.dart';
+import 'package:test/test.dart';
+
+import '../src/context.dart';
+
+/// Matcher for functions that throw [BuildException].
+final Matcher throwsBuildException = throwsA(const isInstanceOf<BuildException>());
+
+void main() {
+  group('BuildAction', () {
+    group('analyze', () {
+      test('throws BuildException if action declares the same file as input and output', () async {
+        await fs.file('a.dart').create();
+        await fs.file('b.dart').create();
+        final BuildAction action = new BuildAction(
+          inputPaths: <String>['a.dart', 'b.dart'].toSet(),
+          outputPaths: <String>['a.dart'].toSet(),
+          buildFunction: (Set<File> inputs) async {},
+        );
+        expect(() async => await action.build(), throwsBuildException);
```

Comment:
You can, though I generally prefer writing `() => x.y()` in `expect` calls just to clarify that it's a method call as opposed to a sync property lookup. That said, it's hard to argue that `() async => await x.y()` is 'readable', so I'm sold -- done!